apps/prairielearn/python/check_data.py

```diff
@@ -0,0 +1,179 @@
+        "edit_phases": frozenset({"grade", "test"}),
+    },
+    "score": {
+        "type": "number",
+        "present_phases": frozenset({"render", "grade", "test"}),
+        "edit_phases": frozenset({"grade", "test"}),
+    },
+    "feedback": {
+        "type": "object",
+        "present_phases": frozenset({"render", "grade", "test"}),
+        "edit_phases": frozenset({"grade", "test"}),
+    },
+    "editable": {
+        "type": "boolean",
+        "present_phases": frozenset({"render"}),
+        "edit_phases": frozenset(),
+    },
+    "manual_grading": {
+        "type": "boolean",
+        "present_phases": frozenset({"render"}),
+        "edit_phases": frozenset(),
+    },
+    "panel": {
+        "type": "string",
+        "present_phases": frozenset({"render"}),
+        "edit_phases": frozenset(),
+    },
+    "num_valid_submissions": {
+        "type": "integer",
+        "present_phases": frozenset({"render"}),
+        "edit_phases": frozenset(),
+    },
+    "gradable": {
+        "type": "boolean",
+        "present_phases": frozenset({"parse", "grade", "test"}),
+        "edit_phases": frozenset(),
+    },
+    "filename": {
+        "type": "string",
+        "present_phases": frozenset({"file"}),
+        "edit_phases": frozenset(),
+    },
+    "test_type": {
+        "type": "string",
+        "present_phases": frozenset({"test"}),
+        "edit_phases": frozenset(),
+    },
+    "extensions": {
+        "type": "object",
+        "present_phases": all_phases,
+        "edit_phases": frozenset(),
+    },
+}
+
+
+def check_prop(
+    prop: str,
+    old_value: Any,
+    new_value: Any,
+    value_type: ValueType,
+    present_phases: frozenset[Phase],
+    edit_phases: frozenset[Phase],
+    phase: Phase,
+) -> None:
+    # Short-circuit if we don't care about this prop in this phase.
+    if phase not in present_phases:
+        return
+
+    # Ensure that the prop is present.
+    if new_value is None:
+        raise ValueError(f'data["{prop}"] is missing')
+
+    # Check the type.
+    if value_type == "integer" and not isinstance(new_value, int):
+        raise ValueError(f'Expected data["{prop}"] to be an integer')
+    elif value_type == "string" and not isinstance(new_value, str):
+        raise ValueError(f'Expected data["{prop}"] to be a string')
+    elif value_type == "number" and not isinstance(new_value, (int, float)):
+        raise ValueError(f'Expected data["{prop}"] to be a number')
+    elif value_type == "boolean" and not isinstance(new_value, bool):
+        raise ValueError(f'Expected data["{prop}"] to be a boolean')
+    elif value_type == "object" and not isinstance(new_value, dict):
+        raise ValueError(f'Expected data["{prop}"] to be an object')
+
+    # Check the value.
+    if phase not in edit_phases and old_value != new_value:
+        raise ValueError(f'data["{prop}"] has been illegally modified')
+
+
+def check_data(old_data: dict, new_data: dict, phase: Phase) -> None:
```

Comment:
```suggestion
def check_data(old_data: dict[str, Any], new_data: dict[str, Any], phase: Phase) -> None:
```
See other comment. I actually think this function will throw an unintended exception if keys are not all strings.

Reply:
Can you elaborate on the exception-throwing thing?

FWIW, this change feels like lying to the typechecker - `data` is modifiable by users and there's absolutely nothing stopping them from using a non-string as a key.

Reply:
> FWIW, this change feels like lying to the typechecker - `data` is modifiable by users and there's absolutely nothing stopping them from using a non-string as a key.

Ahhh that's a good point, I wasn't thinking about this. Especially since validating these dictionaries is the purpose of this function.

As for the exception throwing, the iterable used in the string `.join()` method needs to only consist of strings, or the method will throw a type error. So this will throw an exception:
```python
items = ['a', 'b', 1]
res_str = ''.join(items)
```
So if the dict we're checking has a non-string key, an exception will get thrown once we try to make the error message for the exceptions in this function.

Reply:
What's the right way to handle that, something like this?

```python
''.join(str(val) for val in [1, '2', 'three'])
```

Reply:
Yeah, I would do `''.join(map(str, iterable))`.

Reply:
Changed in b47aeab; I also added a test for good measure.